### PR TITLE
Support setting cuda arch and improve mesh generation

### PR DESCRIPTION
**Pull request by Levi-Armstrong**
Currently the marching cubes mesher return duplicate vertices, zero area triangles ext. This uses vtk directly to populate the mesh from the algorithm then leverages the vtk clean poly data to clean up the mesh.

### CODE REVIEW

**Comment by Levi-Armstrong** [comment]
@schornakj Would you be interested in removing the dependency on PCL? This could be changed to return a VTK mesh instead then this would only depend on VTK.

**Comment by schornakj** [comment]
> Currently the marching cubes mesher return duplicate vertices, zero area triangles ext. This uses vtk directly to populate the mesh from the algorithm then leverages the vtk clean poly data to clean up the mesh.

I think there's an underlying issue with the marching cubes meshing algorithm, since it shouldn't be returning duplicate vertices and non-manifold meshes in the first place. The VTK `cleanPolyData` is a reasonable fix for now, since I know we have downstream mesh processing algorithms that get messed up by zero-area triangles, etc.

> @schornakj Would you be interested in removing the dependency on PCL? This could be changed to return a VTK mesh instead then this would only depend on VTK.

It's convenient for the marching cubes function to return a PCL mesh, since `yak_ros` uses PCL to do a mesh transformation afterwards. This would be an API change so we'd have to be careful about keeping track of version compatibility between `yak`, `yak_ros`, and `yak_ros2`. In principle it's good to simplify package dependencies as much as possible.

**Comment by schornakj** [comment]
@Levi-Armstrong Since you updated the method of triangle list construction to use VTK directly instead of the (possibly buggy) home-rolled method we were using before, do we still have issues with duplicate vertices if we don't do the `cleanPolyData` step?

**Comment by Levi-Armstrong** [comment]
> @Levi-Armstrong Since you updated the method of triangle list construction to use VTK directly instead of the (possibly buggy) home-rolled method we were using before, do we still have issues with duplicate vertices if we don't do the `cleanPolyData` step?

I think it is still an issues. Currently it looks like it just loops over every voxel in the marching cubes and request its triangles and adds them to the list which would produce duplicate vertices. Ideally we would want to get all vertices and then the request for triangles would return vertices indices, but I do not believe the current implementation supports this.

**Comment by Levi-Armstrong** [comment]
> t's convenient for the marching cubes function to return a PCL mesh, since `yak_ros` uses PCL to do a mesh transformation afterwards. This would be an API change so we'd have to be careful about keeping track of version compatibility between `yak`, `yak_ros`, and `yak_ros2`. In principle it's good to simplify package dependencies as much as possible.

I agree, we could mark it deprecated and add another method that returns the vtk mesh type or just support both. PCL provides vtk utilities for converting back and forth.

**Comment by Levi-Armstrong** [comment]
We tested this on an active project and everything appears to work the same.

File: yak/src/mc/marching_cubes.cpp [hunk 1]
```diff
@@ -150,55 +157,61 @@ std::vector<Triangle> processCube(const yak::TSDFContainer& grid, int x, int y,
...
       {
         std::vector<Triangle> ts = processCube(grid, x - 1, y - 1, z - 1, params.min_weight);
-        triangle_buffers[tid].insert(triangle_buffers[tid].end(), ts.begin(), ts.end());
+#pragma omp critical
+        {
+          for (auto& t : ts)
+          {
+            vtkIdType p1 = points->InsertNextPoint(t.v[0].x(), t.v[0].y(), t.v[0].z());
+            vtkIdType p2 = points->InsertNextPoint(t.v[1].x(), t.v[1].y(), t.v[1].z());
+            vtkIdType p3 = points->InsertNextPoint(t.v[2].x(), t.v[2].y(), t.v[2].z());
```
Comment (schornakj) [inline yak/src/mc/marching_cubes.cpp]: These lines produce a warning about implicit conversion from float to double. Can you add some static casts around `t.v[0].x()`, etc.?

**Review by schornakj — CHANGES_REQUESTED** [review]
Looks good! Just one thing to change to avoid an implicit float/double cast.

On my system adding the VTK mesh clean step adds about 1 second to the time needed to mesh the bunny demo (13.7 seconds with cleaning vs. 14.7 seconds without cleaning), which is a minor (and optional) addition to processing time.

File: yak/src/mc/marching_cubes.cpp (hunk 1 above)
Comment (Levi-Armstrong) [reply yak/src/mc/marching_cubes.cpp]: Changes have been made.

**Review by schornakj — APPROVED** [review]

**Pull request merged**